Improve `middle(::AbstractRange)` performance

vyu opened the pull request:
Currently, the behavior of `middle` is inconsistent between ranges and vectors with non-`Real` eltypes:
```julia
julia> using Statistics

julia> r = LinRange(1im, 2im, 2)
2-element LinRange{ComplexF64, Int64}:
 0.0+1.0im,0.0+2.0im

julia> middle(r)
0.0 + 1.5im

julia> middle(collect(r))
ERROR: MethodError: no method matching isless(::ComplexF64, ::ComplexF64)
[...]
```

This PR fixes this by restricting `middle(::AbstractRange)` to `Real` eltypes. Performance is improved by calling `mean`, which elides bounds checks and does not call `length` (which is slow for `StepRange`).

Also, the doc for `middle(range)` is incorrect since ranges are not always sorted (there might not be a canonical order on the eltype, as above), so I've removed it and moved its example to the `middle(array)` doc.

vyu commented on src/Statistics.jl:
```diff
@@ -801,6 +789,11 @@ julia> middle(a)
 """
 middle(a::AbstractArray) = ((v1, v2) = extrema(a); middle(v1, v2))
 
+function middle(a::AbstractRange{<:Real})
```
Ah, I see. I've reverted the restriction to real eltypes, so it now preserves previous behavior and this PR just improves `middle`'s performance by calling `mean`. How does it look now?

nalimilan commented on src/Statistics.jl:
```diff
@@ -764,21 +764,6 @@ equivalent in both value and type to computing their mean (`(x + y) / 2`).
-middle(a::AbstractRange) = middle(a[1], a[end])
-
 """
     middle(a)
```
While we're at it:
```suggestion
    middle(a::AbstractArray)
```

nalimilan commented on src/Statistics.jl:
```diff
@@ -801,6 +789,11 @@ julia> middle(a)
 
+function middle(a::AbstractRange)
+    isempty(a) && throw(ArgumentError("middle of an empty range is undefined."))
+    return mean(a)
```
Given that `mean` only does this (plus a redundant check), better just copy the definition here:
```suggestion
    return (first(r) + last(r)) / 2
```

nalimilan commented on test/runtests.jl:
```diff
@@ -21,6 +21,9 @@ Random.seed!(123)
         @test middle(one(T)) === middle(one(T), one(T))
     end
+
+    @test_throws Exception middle(Int[])
+    @test_throws Exception middle(1:0)
```
```suggestion
    @test_throws MethodError middle(Int[])
    @test_throws ArgumentError middle(1:0)
```

Can you also add a test covering the behavior you aimed to fix in this PR?

vyu replied:
I was thinking that testing for a generic `Exception` here would be more semantically appropriate because the fact that empty vectors and ranges return different exceptions is incidental rather than intentional. But I don't mind changing to the concrete exceptions if you prefer that.

vyu replied:
Okay, makes sense. Changed.

nalimilan reviewed (commented):
Thanks! Though I don't understand what bug you want to fix here. Could you post an example (or add a test) of something that didn't work (or worked incorrectly) but does with the PR?

ararslan commented:
> Fix middle for non-1-based ranges as @ararslan pointed out, but the `AbstractRanges` in `Base` are all 1-based so I don't have a simple test for this.

Who needs Base anyway? 😄 

```julia
struct WeirdRange{T} <: AbstractRange{T}
    r::UnitRange{T}
end

Base.firstindex(wr::WeirdRange) = -2

Base.lastindex(wr::WeirdRange) = firstindex(wr) + length(wr) - 1

Base.length(wr::WeirdRange) = length(wr.r)

Base.axes(wr::WeirdRange) = (firstindex(wr):lastindex(wr),)

Base.step(wr::WeirdRange) = step(wr.r)

function Base.getindex(wr::WeirdRange, i::Integer)
    i in eachindex(wr) || throw(BoundsError(wr, i))
    return wr.r[i - firstindex(wr) + 1]
end

wr = WeirdRange(0:3)
middle(wr[1], wr[end])  # WRONG: returns 3.0
mean(wr)                # CORRECT: returns 1.5
```

nalimilan commented:
OK. Not sure it's worth testing non-1 based ranges if they don't exist anywhere TBH... But `middle(0:typemax(Int))` sounds easy enough to test.

vyu commented:
Okay, I've added a test for `middle(0:typemax(Int))`. Also fixed a CI failure on 1.7 due to a change of exception type in 1.8 by adding a version check.

ararslan reviewed (approved):
Great work!
